### PR TITLE
Shared AzureKeyCredential

### PR DESCRIPTION
**Pull request by maririos**
- Common `AzureKeyCredential` as part of `Azure.Core`
- Common `AzureKeyCredentialPolicy` as a shared source

Changes reflected in Text Analytics and Search as starting point.

Note that I had to add `UseProjectReferenceToAzureClients` to both projects to do P2P instead of package reference. Once we release `Azure.Core` we can remove this.

~Creating as draft to gather feedback on approach~

### CODE REVIEW

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Is this because the policy is shared source?

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: Add a simple test for it.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Maybe we should just ship it as public API.

Comment (schaabs) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: This property should be called Key not ApiKey

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (schaabs) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: The parameter should be called key vs. apiKey

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (schaabs) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: The parameter should be called key vs. apiKey

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (tg-msft) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: I wonder if we want to put this in the main `Azure` namespace.  `TokenCredential` is in `Azure.Core` but most folks aren't meant to use that directly.  When I think about the getting started experience for Search, TextAnalytics, etc., it'd be nice to not have to add an extra `using Azure.Core;`.  Let's leave it here for this PR and follow up with @KrzysztofCwalina on whether to move it before the next Preview.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (tg-msft) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: We've been kind of lazy in how we hide these properties on other credential types.  There's no meaningful security benefit, but we're trying to guide users to not rely on reading keys from our types and instead be thoughtful about how they manage secure configuration details in their application.  I think the `EditorBrowsable` approach here is a clever way of walking that fine line by keeping our scenarios possible while still hiding it from the type of dev who might be tempted to misuse it.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Also, would it be better to call the method UpdateKey?

**Review by tg-msft — APPROVED** [review]
This looks awesome - thanks @maririos for updating Search everywhere.   I'd love to get this checked in tomorrow if possible because we're looking at [renaming Search for reasons](https://github.com/Azure/azure-sdk-for-net/issues/10634) and I don't want to cause you any avoidable merge pain.

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (KrzysztofCwalina) [inline sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: @schaabs, are there cases where the key needs to be somehow transformed before it ends up in the header or it's written to the header as-is?

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (KrzysztofCwalina) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: @pakrym, should it be here or in Azure root?

Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: I somewhat agree with https://github.com/Azure/azure-sdk-for-net/pull/10668#discussion_r394031804 but for consistency I would've kept it in Azure.Core.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: We need to decide if we want it public or not, if it is we shouldn't have a method that's basically a setter for the same property.

Do we have a scenario where this property is worth hiding? Because people who tend to missuse things usually find ways to do it and I doubt EditorBrowsable would stop them.

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (schaabs) [reply sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: There are cases where the key is not directly inserted as a header value, but rather used to sign a request. However this policy is specifically for cases where the key is simply added as a header. Perhaps a more specific type name would be better, but given that this type is internal it would be for library authors benefit.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (schaabs) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: I agree `UpdateKey` is preferable to `UpdateCredential`. I would even say it could be called simply `Update`, but I would defer to @KrzysztofCwalina  on that.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Everyone should be using DefaultAzureCredential anyway

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Why not turn it a property setter?

Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Update works for me too.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: I just worry that if we keep it in Azure.Core, people will start adding Azure.Core namespace to their files. I think this type is simple enough and will be used in many services, and so Azure root  would make the experience better.

Comment (schaabs) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: @pakrym I don't believe it's the same case as TokenCredential. Perhaps TokenCredential should have been in Azure rather than Azure.Core, however a library user will very infrequently use the type. They will more likely use an implementation from Azure.Identity like DefaultAzureCredential. However this type they will have to reference directly.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: I like Update because it can be used across different credential types to roll the credential.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (AlexanderSher) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Do we need this method at all?

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (AlexanderSher) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Is it allowed to assign `null` to `_key` here?

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Updated

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Update works for me too.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Interesting. I will leave it here for this PR, and will revisit the namespace.
Issue https://github.com/Azure/azure-sdk-for-net/issues/10672

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Removed

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: Good point. Will do

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: >we shouldn't have a method that's basically a setter for the same property.

You mean the `Update` method? We want to be able to rotate the key without restarting clients, which is why we have the method available.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: The only way to set value to property is in the `Update` method which has validation for that.
https://github.com/Azure/azure-sdk-for-net/pull/10668/files#diff-dca2442a6e0ae3f87d86879dc5dc2afbR43

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Right, but having a public setter on the property would accomplish the same thing.

Comment (tg-msft) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: A property setter would work here, but not for a lot of other credential types.  An `Update` method works better if you need to do any computation on the values, if you have multiple values that need to be swapped atomically, etc.  This is more about following @schaabs's Credential pattern even though the `Update` method isn't as necessary with this simplest of credentials.

Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: True. The [guidelines](https://azure.github.io/azure-sdk/general_design.html#authentication) say:
> DO define an update method which accepts all mutable credential data, and updates the credential in an atomic, thread safe manner.

So following just in case later we need to make modifications to the key

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: Test added

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Issue #10672

Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Yeah, I would put it in the root.

**Comment by pakrym** [comment]
Can you update Azure.Core changelog please?

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Will do

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (heaths) [inline sdk/core/Azure.Core/CHANGELOG.md]: Should be under an `### Added` heading according to [guidelines](https://azure.github.io/azure-sdk/policies_releases.html#changelog-guidance).

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (heaths) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Document exceptions.

Comment (heaths) [reply sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Should also put on the constructor since it calls this method.

File: sdk/core/Azure.Core/src/AzureKeyCredential.cs
Comment (heaths) [inline sdk/core/Azure.Core/src/AzureKeyCredential.cs]: Tip: not wrong, but you could just mark the field as `volatile`.

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (heaths) [inline sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: Document exceptions, though since this is internal it's not as big a deal. Definitely should for public APIs, though (documentation to set expectations).

**Review by heaths — APPROVED** [review]
General and Search changes LGTM, with a few comments to consider.

File: sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs
Comment (johanste) [inline sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs]: Depending on how general purpose you want to make the class, some services (although I'm not sure if any Azure services that do this) have multiple different API keys, any one of which could be used. In practical terms, it means that depending on which kind of API key is provided, different headers are used (e.g. an AppId API key is sent in the X-APPID-KEY header, and a user API key is sent in the X-USER-KEY header). 

I guess you could just subclass the key in a service specific key credential at that point, though.

Comment (maririos) [reply sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs]: Correct me if I am wrong, but this scenarios is covered with the policy, as it currently allows callers to set the header of the key they are going to use: https://github.com/Azure/azure-sdk-for-net/pull/10668/files#diff-60cd6c136a6b3ad6e0dcfd38c41cb72fR18

Comment (johanste) [reply sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs]: Maybe. The question would be how a user of one of our client libraries indicate that the specific ApiKeyCredential they have is a user api key vs. app api key in the example above. If there is only one type of api key, the policy can have that knowledge baked in. If there are two api key types (each with it's own corresponding header or query parameter) for the same service, someone is going to have to help the policy figure out which header to use.

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (johanste) [inline sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: Api keys can sometimes be passed as query parameters. One such example is Azure cognitive search. (however, for azure cognitive search, you can use both a header or a query parameter interchangeable, so we can pick one). 

I don't know if we currently have services that only support query string api keys in Azure, but we had a discussion this week for a service that was considering either that or a shared access signature query parameter.

File: sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs
Comment (maririos) [reply sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs]: Thanks Johan! Yeah, definitely something to consider.
For now I am going to leave it as is, and will keep evaluating for those future scenarios

File: sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/Shared/AzureKeyCredentialPolicy.cs]: For .NET, we are keeping this policy in a private API via shared source. People can use it or create their own logic for the policy in order to handle that scenario.
Will need to see how this will work for other languages though

**Comment by maririos** [comment]
Merging for now to unblock Ted. Will keep addressing comments and create future PRs as needed

**Pull request merged**
